pkg/devicetwin/devicetwin_test.go

```diff
@@ -0,0 +1,182 @@
+	fakeModule = beehive.NewMockModule(mockCtrl)
+	dbm.DBAccess = ormerMock
+}
+
+// registerFakeModule is fuction to register all mock modules used.
+func registerFakeModule() {
+	fakeModule.EXPECT().Name().Return(TestModule).Times(3)
+	core.Register(fakeModule)
+	mainContext = context.GetContext(context.MsgCtxTypeChannel)
+	mainContext.AddModule(TestModule)
+}
+
+// TestName is function to test Name().
+func TestName(t *testing.T) {
+	tests := []struct {
+		name string
+		want string
+	}{
+		{
+			name: "DeviceTwinNametest",
+			want: "twin",
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			dt := &DeviceTwin{}
+			if got := dt.Name(); got != tt.want {
+				t.Errorf("DeviceTwin.Name() = %v, want %v", got, tt.want)
+			}
+		})
+	}
+}
+
+// TestGroup is function to test Group().
+func TestGroup(t *testing.T) {
+	tests := []struct {
+		name string
+		want string
+	}{
+		{
+			name: "DeviceTwinGroupTest",
+			want: "twin",
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			dt := &DeviceTwin{}
+			if got := dt.Group(); got != tt.want {
+				t.Errorf("DeviceTwin.Group() = %v, want %v", got, tt.want)
+			}
+		})
+	}
+}
+
+// TestStart is function to test Start().
+func TestStart(t *testing.T) {
+	initMocks(t)
+	registerFakeModule()
+	core.Register(&DeviceTwin{})
+	dt := DeviceTwin{}
+	mainContext.AddModule(dt.Name())
+	mainContext.AddModuleGroup(dt.Name(), dt.Group())
+	dt.context = mainContext
+	controller, err := InitDTController(mainContext)
+	if err != nil {
+		t.Errorf("Start device twin failed, due to %v", err)
+	}
+	dt.dtcontroller = controller
+	ormerMock.EXPECT().QueryTable(gomock.Any()).Return(querySeterMock).Times(200)
+	querySeterMock.EXPECT().All(gomock.Any()).Return(int64(1), nil).Times(200)
+	go dt.Start(mainContext)
+	// Sending a message from devicetwin module to the created fake module(TestModule).
+	msg := model.NewMessage("").BuildRouter(TestModule, GroupResource, model.ResourceTypePodStatus, model.InsertOperation)
+	dt.context.Send(TestModule, *msg)
+	_, err = mainContext.Receive(TestModule)
+	t.Run("MessagePingTest", func(t *testing.T) {
+		if err != nil {
+			t.Errorf("Error while receiving message: %v", err)
+			return
+		}
+	})
+	//Checking health of Mem,Twin,Device and Comm module
+	tests := []struct {
+		name       string
+		moduleName string
+	}{
+		{
+			name:       "MemModulehealthCheck",
+			moduleName: dtcommon.MemModule,
+		},
+		{
+			name:       "TwinModulehealthCheck",
+			moduleName: dtcommon.MemModule,
+		},
+		{
+			name:       "DeviceModulehealthCheck",
+			moduleName: dtcommon.MemModule,
+		},
+		{
+			name:       "CommModulehealthCheck",
+			moduleName: dtcommon.MemModule,
+		},
+	}
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			health, ok := controller.DTContexts.ModulesHealth.Load(test.moduleName)
+			if ok {
+				now := time.Now().Unix()
+				if now-health.(int64) > 60*2 {
```

Comment:
in the code of dtcontroller start , if the modules dont ping / send heartbeat after more than 60*2 seconds , then  it is assumed that that module is unhealthy and it will attempt to start it again. so we dont need to check that case here .
tests could be like : 

1. after dtcontroller starts , it registers these modules first , so we can check if the context does have all these 4 modules.
2. get the heartbeat from the module health for each of these modules and see if ok. if not ok , just log a debug statement , saying that the heartbeat may have expired for this module and will be handled by dtcontroller to start it again.